### PR TITLE
Adding GitHub Enterprise as a source provider

### PR DESCRIPTION
**Pull request by jpricketMSFT**
I also did a bit of refactoring since the implementation for GitHub, Bitbucket, and GitHub Enterprise were all identical except for one line.

### CODE REVIEW

File: src/Agent.Worker/Build/GitSourceProvider.cs [hunk 1]
```diff
@@ -30,16 +30,14 @@ public override string GenerateAuthHeader(string username, string password)
             throw new NotSupportedException(nameof(ExternalGitSourceProvider.GenerateAuthHeader));
         }
     }
-
-    public sealed class GitHubSourceProvider : GitSourceProvider
+
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/GitSourceProvider.cs]: extra space. :)

File: src/Agent.Worker/Build/GitSourceProvider.cs [hunk 2]
```diff
@@ -30,16 +30,14 @@ public override string GenerateAuthHeader(string username, string password)
             throw new NotSupportedException(nameof(ExternalGitSourceProvider.GenerateAuthHeader));
         }
     }
-
-    public sealed class GitHubSourceProvider : GitSourceProvider
+    
+   public abstract class AuthenticatedGitSourceProvider : GitSourceProvider
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/GitSourceProvider.cs]: missing 1 space.

**Comment by TingluoHuang** [comment]
Looks good.

**Comment by TingluoHuang** [comment]
@scottboehmer FYI.

**Comment by jpricketMSFT** [comment]
@TingluoHuang - Thanks Ting!

**Pull request merged**